### PR TITLE
Add Cognito User Pools clients

### PR DESCRIPTION
**Pull request by c0va23**
Hi, I have the intention to add support for Cognito Identity Provider API clients to this set of extensions. I am using Quarkus on the current project and in the near future I will implement user management, which we store in the Cognito User Pool.

My proposed implementation of the extension for Cognito User Pools I did by analogy with the current implementation of the extension for SQS.

I have successfully tested the injection and configuration of synchronous and asynchronous clients. And build a native image with a new extension. I checked the steps from the documentation locally, except for generating a project with the extension connected.

I have now implemented the following:
- Synchronous and asynchronous client configuration
- Provider for synchronous and asynchronous client
- Two unit tests for configurations (valid and invalid)
- Simple integration test
- Generation of documentation for the configuration
- Wrote simple documentation

I added `motoserver` to the integration tests because local-stack supports Cognito only in the Pro version.

Do I have a chance of accepting these changes? And what else do I need to do to increase my chances?

### CODE REVIEW

File: README.md
Comment (gsmet) [inline README.md:21]: Maybe add it at the beginning? We kept the alphabetical order here.

File: bom/pom.xml
Comment (gsmet) [inline bom/pom.xml:127]:
```suggestion
                <artifactId>quarkus-amazon-cognito-idp</artifactId>
```
To be more consistent with the rest, I would just call the extension `quarkus-amazon-cognito`. AFAICS, Cognito is just that right? Or you expect it to require some additional extensions later?

**Review by gsmet — COMMENTED** [review]
Interesting and very complete work, thanks! Looks typically like something we would want in.

Can you have a look at my suggestion and question?

File: bom/pom.xml
Comment (c0va23) [reply bom/pom.xml:127]: Cognito have three API (https://docs.aws.amazon.com/cognito/index.html). These APIs do not have stable names (Cognito Sync has no such problem).

| Docs name | SDK package  | aws-cli subcommand |
|-----------------|---------------------|-----------------------------|
| [Amazon Cognito user pools](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools.html) | [cognitoidentityprovider](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/cognitoidentityprovider/package-frame.html) | [`cognito-idp`](https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/index.html) |
| [Amazon Cognito identity pools](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-identity.html) | [cognitoidentity](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/cognitoidentity/package-frame.html) | [`cognito-identity`](https://docs.aws.amazon.com/cli/latest/reference/cognito-identity/index.html) |

I chose the name `cognito-idp` as the shortest and most readable.

File: README.md
Comment (c0va23) [reply README.md]: Fixed

**Comment by gsmet** [comment]
@kdubb any chance you could also have a look at this one?

**Comment by kdubb** [comment]
@gsmet Will do. I am a little behind at the moment, so give me some time.

File: bom/pom.xml
Comment (kdubb) [reply bom/pom.xml:127]: @c0va23 The naming situation is unfortunate. I must say that after looking at the linked website, when I see "cognito-idp" I think it's for "identity pools".

My suggestion is to use `cognito-user-pools`, `cognito-identity-pools` and `cognito-sync`.

Comment (c0va23) [reply bom/pom.xml]: @kdubb, i rename extension to `cognito-user-pools`.

File: cognito-user-pools/runtime/src/main/java/io/quarkus/amazon/cognitouserpools/runtime/CognitoUserPoolsClientProducer.java
Comment (kdubb) [inline cognito-user-pools/runtime/src/main/java/io/quarkus/amazon/cognitouserpools/runtime/CognitoUserPoolsClientProducer.java:39]: Message should reference `CognitoIdentityProviderAsyncClient`

File: docs/modules/ROOT/pages/amazon-cognitouserpools.adoc
Comment (kdubb) [inline docs/modules/ROOT/pages/amazon-cognitouserpools.adoc:70]: Odd grammar. Maybe `The application built here shows the count of Cognito User Pools in the current region.`

**Review by kdubb — CHANGES_REQUESTED** [review]
Overall looks good and very consistent with the other API implementations. Just a couple small things to fix.

**Comment by c0va23** [comment]
@kdubb, i corrected the comments. Made a rebase on main. And corrected the name of the extension in several places.

File: cognito-user-pools/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (kdubb) [inline cognito-user-pools/runtime/src/main/resources/META-INF/quarkus-extension.yaml:3]: Shouldn't the name include "user pools"?

File: cognito-user-pools/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (kdubb) [inline cognito-user-pools/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: Same thing for keywords. Shouldn't they include "user pools"?

**Review by kdubb — COMMENTED** [review]
Missed these on the first review, this should be it though.

File: cognito-user-pools/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (c0va23) [reply cognito-user-pools/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: Fixed

File: cognito-user-pools/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (c0va23) [reply cognito-user-pools/runtime/src/main/resources/META-INF/quarkus-extension.yaml:10]: Added "congito-user-pools"

**Review by kdubb — APPROVED** [review]

**Comment by emattheis** [comment]
@gsmet @kdubb Any reason this can't get merged at this point? I just started doing some work with cognito and would love to have this available in Quarkus.

**Comment by emattheis** [comment]
@c0va23 I just pulled this to my local fork to check it out and I noticed that the new extension name doesn't follow the established naming convention. The other extension names are aligned with the names used in the AWS CLI for the respective service - e.g. `quarkus-amazon-sqs` provides capabilities to `$ aws sqs <command>` - so this extension should be `quarkus-amazon-cognito-idp` instead of `quarkus-amazon-cognito-user-pools`. I thought this convention was documented somewhere, but I can't seem to find it anymore...

**Comment by kdubb** [comment]
@emattheis I requested the "user-pools" name. No, it may not align with the AWS CLI but I believe it does align better with the AWS SDK naming and documentation. This implementation is only about user pools and not identity pools and I believe the extension name should reflect that. Using the `idp` name from CLI naming really implies that this implementation is about identity pools.

**Comment by emattheis** [comment]
@kdubb I see. While I agree that the AWS naming isn't fantastic (they have `cognito-idp` for user pools and `cognito-identity` for identity pools), but alignment with the AWS CLI and Java SDK is very useful. It makes it immediately obvious which extensions match the respective AWS API. We're effectively just manually generating boilerplate wrapper code and documentation for the generated API clients that Amazon maintains.

**Comment by emattheis** [comment]
Just looked at the resolved comments. I wish I was paying attention to the repo earlier in the year 😛 

It's important to keep in mind that AWS Cognito User Pools is in fact an OAuth2 IDP, as opposed to AWS Identity Pools which is a way to apply RBAC to identiies (possibly provided by user pools).

So, while the confusion between the respective AWS offerings is certainly real, I don't think it's something we should try to address in Quarkus.

**Comment by kdubb** [comment]
@emattheis Where in the Java SDK is this referred to as `idp`? That appears to only be AWS CLI command naming and I believe that actually refers to identity provider. The Java SDK has them in the `software.amazon.awssdk.services.cognitoidentityprovider` package and refers to them as `UserPools`.

**Comment by emattheis** [comment]
@kdubb `cognito-idp` is only used in the CLI AFAIK, but it correlates directly with https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/cognitoidentityprovider/package-summary.html, whereas `cognito-identity` in the CLI correlates to https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/cognitoidentity/package-summary.html.

**Comment by kdubb** [comment]
Yeah AWS really dropped the ball on naming here 🤦‍♂️

I don't see the correlation you are referring to between the CLI and SDK javadoc but at this point we're really just splitting hairs.

In summary, my vote is to include "user pools" in the name because that aligns directly with the Java SDK naming as well as AWS's online documentation terminology and it is the only API this extension supports. If I end up being outvoted I think I'll be ok haha.

**Comment by emattheis** [comment]
The correlation is between the APIs covered by the CLI and the Java SDK. Every action you can perform with `aws cognito-idp` can be done via `software.amazon.awssdk.services.cognitoidentityprovider.CognitoIdentityProviderClient` or `software.amazon.awssdk.services.cognitoidentityprovider.CognitoIdentityProviderAsyncClient`.

For all the current extensions, you can take the CLI command and prefix it with `quarkus-amazon-` to find the equivalent extension. Likewise, a service provided by the AWS Java SDK v2 at `software.amazon.awssdk.services.<SERVICE_NAME>` is made available by a Quarkus extension at `io.quarkus.amazon.<SERVICE_NAME>`.

My vote is for consistent alignment with Amazon so we don't have to think about it, but it's not the hill I want to die on, either.

(Note that I am also working on an extension for sending messages to websocket connections managed by AWS API Gateway. The absurd name for the AWS API involved is `apigatewaymanagementapi`, but at least the package name and the CLI name are aligned 🤦)

**Comment by gsmet** [comment]
@c0va23 sorry, was busy on other things. I think we should proceed with merging this: only problem is that CI has not run on the PR for whatever reason. I tried to rebase it but I can't push to your fork. Could you rebase it and ping me when done? I will merge it once CI is green.

**Comment by gsmet** [comment]
Also please do a proper rebase (i.e. avoid having a merge commit). Thanks!

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Note: if we get this merged by Monday, it will be included in the Quarkus 2.8.0.Final Platform release.

**Comment by emattheis** [comment]
@gsmet thoughts on the naming discussion I had with @kdubb yesterday? Not sure who else should weigh in, but I think we should be careful about deviating from the existing convention. I'm fine being outvoted, but I could have sworn the alignment with Amazon naming was documented at one point. Would be nice if some of the original authors could weigh in...

**Comment by c0va23** [comment]
@gsmet , I did a rebase and re-generated the documentation.

**Comment by gsmet** [comment]
Thanks and sorry for the delay!

**Pull request merged**
